### ISSUE TITLE
Elasticsearch 2.0 beta testing: Performance difference between similar aggs vs. query

### ISSUE THREAD

**Issue by spinscale** [elastic/elasticsearch#13964]
TLDR: Searches with sorting on a field to get the min/max value of a dataset appears to be a bit faster than aggregating using `min/max` aggs. Maybe just a flaw in my test, feel free to point it out!

While testing 2.0, I did a simple test. Using `makelogs` to index 2 million documents

``` bash
makelogs -c 2m -d 10 --verbose --omit relatedContent
```

Note: You need to install the latest makelogs if you want to use it: `npm install -g makelogs@beta`

Now I executed the following two queries, thinking that they did not have any difference in execution speed

``` json
GET logstash-*/_search
{
  "size": 1, 
  "sort": [
    {
      "@timestamp": {
        "order": "desc"
      }
    }
  ]
}

GET logstash-*/_search?search_type=count
{
  "aggs": {
    "foo": {
      "max": {
        "field": "@timestamp"
      }
    }
  }
}
```

However, the search seems to be a bit faster. I gathered data by running this and taking the `took` time from the responses, including one warmup round and no indexing happening:

`````` bash
for i in {1..500} ; do curl -s 'localhost:9200/logstash-*/_search' -d '{ "size": 1, "sort": [ { "@timestamp": { "order": "desc" } } ] }' | underscore extract 'took'   ; done | tee -a took-search.txt
for i in {1..500} ; do curl -s 'localhost:9200/logstash-*/_search?search_type=count' -d '{ "aggs": { "foo": { "max": { "field": "@timestamp" } } } }' | underscore extract 'took'   ; done | tee -a took-aggs.txt```

Those are the results

```bash
$ cat ../took-search.txt| st --complete
N     min   q1    median  q3  max   sum     mean    stddev    stderr
500   18    24.5  26      28  120     13524   27.048  6.50156     0.290758

$ cat ../took-aggs.txt| st --complete
N     min   q1    median  q3  max   sum     mean    stddev    stderr
500   26    33    35      37  57    17643   35.286  3.77528   0.168835

$ cat ../took-search.txt| ./data_hacks/histogram.py -m 18 -x 50
# NumSamples = 500; Min = 18.00; Max = 50.00
#2 values outside of min/max
# Mean = 27.048000; Variance = 42.185696; SD = 6.495052; Median 26.000000
# each ∎ represents a count of 2
   18.0000 -    21.2000 [    12]: ∎∎∎∎∎∎
   21.2000 -    24.4000 [   113]: ∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎
   24.4000 -    27.6000 [   217]: ∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎
   27.6000 -    30.8000 [    99]: ∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎
   30.8000 -    34.0000 [    38]: ∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎
   34.0000 -    37.2000 [     8]: ∎∎∎∎
   37.2000 -    40.4000 [     3]: ∎
   40.4000 -    43.6000 [     5]: ∎∎
   43.6000 -    46.8000 [     3]: ∎
   46.8000 -    50.0000 [     0]:

$ cat ../took-aggs.txt| ./data_hacks/histogram.py -m 18 -x 50
# NumSamples = 500; Min = 18.00; Max = 50.00
#4 values outside of min/max
# Mean = 35.286000; Variance = 14.224204; SD = 3.771499; Median 35.000000
# each ∎ represents a count of 2
   18.0000 -    21.2000 [     0]:
   21.2000 -    24.4000 [     0]:
   24.4000 -    27.6000 [     6]: ∎∎∎
   27.6000 -    30.8000 [    27]: ∎∎∎∎∎∎∎∎∎∎∎∎∎
   30.8000 -    34.0000 [   175]: ∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎
   34.0000 -    37.2000 [   188]: ∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎
   37.2000 -    40.4000 [    75]: ∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎
   40.4000 -    43.6000 [    16]: ∎∎∎∎∎∎∎∎
   43.6000 -    46.8000 [     4]: ∎∎
   46.8000 -    50.0000 [     5]: ∎∎

# Calculate the 95% time
$ cat ../took-search.txt| ./data_hacks/ninety_five_percent.py
34   
$ cat ../took-aggs.txt| ./data_hacks/ninety_five_percent.py
41
``````

Tools used here: [st](https://github.com/nferraz/st) and [data_hacks](https://github.com/bitly/data_hacks)

The performance difference of those two queries is actually bigger on 1.7.

If there are flaws in my measurements, please close immediately, cant rule it out.

All of these tests happened on my MBP on osx.

**Comment by javanna** [comment]
@spinscale is there anything left to do here?

**Comment by spinscale** [comment]
yes, closing the issue.
